lib/utils/rename.js

```diff
@@ -1,16 +1,182 @@
 'use strict'
 var fs = require('graceful-fs')
+var path = require('path')
+var gentlyRm = require('./gently-rm')
 var SaveStack = require('./save-stack.js')
 
 module.exports = rename
 
 function rename (from, to, cb) {
   var saved = new SaveStack(rename)
-  fs.rename(from, to, function (er) {
-    if (er) {
-      return cb(saved.completeWith(er))
-    } else {
-      return cb()
+
+  var basePath = process.cwd()
+  var currentPath = path.resolve(basePath, from)
+  var targetPath = path.resolve(basePath, to)
+
+  var hadErr = false
+
+  var started = 1 // Unlike ncp we always start with 1 (the top-level rename)
+  var finished = 0
+  var running = 0
+  var limit = 512
+
+  // First, try a 'standard' rename on the top-level directory
+  fs.rename(currentPath, targetPath, function (renameErr) {
+    if (!renameErr) {
+      return cb(null)
     }
+
+    if (renameErr.code !== 'EXDEV' && renameErr.code !== 'EPERM') {
+      return cb(saved.completeWith(renameErr))
+    }
+
+    getStats(currentPath)
   })
+
+  function startRenameOrCopy (source) {
+    if (hadErr) return
+
+    var target = source.replace(currentPath, targetPath)
+    started++
+
+    // Always try to rename first
+    fs.rename(source, target, function (err) {
+      if (!err) {
+        return doneOne(true)
+      }
+
+      return getStats(source)
+    })
+  }
+
+  function getStats (source) {
```

Comment:
This function name seems a bit too centered on the mechanics and not the logic of what it's doing. Which is… looking at what kind of thing the path is and selecting the appropriate path for renameOrCopy?